codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,85 +315,96 @@ private Http2Stream requireStream(int streamId) {
+                    if (entry instanceof ByteBuf) {
+                        ByteBuf buf = (ByteBuf) entry;
+                        if (buf.readableBytes() > frameAllowance) {
+                            ByteBuf slice = buf.readSlice(frameAllowance);
+                            toWrite.addComponent(slice);
+                            toWrite.writerIndex(toWrite.writerIndex() + slice.readableBytes());
+                            // The remaining buffer is kept as a retained slice of the original
+                            bufsAndPromises.addFirst(buf.slice().retain());
+                            frameAllowance = 0;
+                            break;
+                        } else {
+                            toWrite.addComponent(buf);
+                            toWrite.writerIndex(toWrite.writerIndex() + buf.readableBytes());
+                            frameAllowance -= buf.readableBytes();
+                        }
                     } else {
-                        // Create a new promise and listen to it for failure
-                        writePromise = ctx.newPromise();
-                        writePromise.addListener(this);
-                    }
-                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
-                        // If we're writing a subset of a composite buffer then we want to release
-                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
-                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
-                        // 0.
-                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
-                        writePromise.addListener(new ChannelFutureListener() {
+                        final ChannelPromise nested = (ChannelPromise) entry;
+                        promise.addListener(new ChannelFutureListener() {
                             @Override
                             public void operationComplete(ChannelFuture future) throws Exception {
-                                toFree.discardReadComponents();
+                                if (future.isSuccess()) {
+                                    nested.trySuccess();
+                                } else {
+                                    nested.tryFailure(future.cause());
+                                }
                             }
                         });
                     }
-                    frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                            size == bytesWritten && endOfStream, writePromise);
-                } while (size != bytesWritten && allowedBytes > bytesWritten);
-            } finally {
-                size -= bytesWritten;
-            }
+                }
+
+                int writeablePadding = Math.min(frameAllowance, padding);
```

Comment:
This is not a new issue but with our current scheme we could end up in sending `n` data frames with all of the data followed by 1 (or `m`) data frames with only padding.  No action here and I think we already have an issue for this though https://github.com/netty/netty/issues/2807....just reminded me that maybe we should re-visit and see how others maybe handling this situation :)